File: 153. Find Minimum in Rotated Sorted Array/153_find-minimum-in-rotated-sorted-array.java
class FindMin {
 public int findMin(int[] nums) {
     if (nums == null || nums.length <= 0) {
         return 0;
     }
     
     int len = nums.length;
     if (len == 1) {
         return nums[0];
     }
     
     int start = 0;
     int end = nums.length - 1;
     while (start < end) {
         int mid = (start + end) >> 1;
         // 需要与nums[end]比较才能判断出来mid的位置
         if (nums[mid] > nums[end]) {
             start = mid + 1;
         } else if (nums[mid] < nums[end]) {
             end = mid;
         } else {
             // 存在重复元素时，此题目不重复 可直接使用break
             end = end - 1;
         }
     }
     return nums[start];
 }
}

File: 153. Find Minimum in Rotated Sorted Array/153_find-minimum-in-rotated-sorted-array.py

class Solution {
public:
    int findMin(vector<int>& nums) {
        if(nums.empty())
            return -1;

        int index1 = 0;
        int index2 = nums.size()-1;
        int IndMid = index1;

        while(nums[index1] >= nums[index2])
        {
            if(index2 - index1 == 1)
            {
                IndMid = index2;
                break;
            }

            IndMid = index1 + (index2-index1)/2;
            //IndMid = (index2 + index1) /2;

            if(nums[index1] == nums[index2] && nums[index1] == nums[IndMid])
                return MinInOrder(nums, index1 ,index2);
            if(nums[index1] < nums[IndMid])
                index1 = IndMid;
            else if(nums[index2] > nums[IndMid])
                index2 = IndMid;
        }
        return nums[IndMid];
    }

    int MinInOrder(vector<int>& nums, int left, int right)
    {
        int result = nums[left];

        for(unsigned int i = left+1; i < right; i++)
        {
            if(result > nums[i])
                result = nums[i];
        }

        return result;
    }
};
